### ISSUE TITLE
Improve Device->Host and Host->Device transfer speed

### ISSUE THREAD

**Issue by pavlepopovic** [tenstorrent/tt-metal#20518]
In some CNN models (usually UNET-like ones), output is an image, which needs to be transferred back to the host.
In an usual setup, this can slow down time between two inferences, as the output tensor needs to be transferred back to the host, before the next inference can begin.
Based on my measurements, on a x86 CPU machine with x8 pcie gen4, I am getting output read BW of about 1.8-3 GiB/s (depending on the host, and tensor size), which is quite off the theoretical 16 GiB/s based on pcie speed.
Now, this is not an immediate priority, as we can do the following:
- Offload tensor to Dram
- Put CQ1 to (in addition to writing inputs, also read outputs)
But still would be good to speed this up.
For reference, input writes seem much faster, and I was able to get 4.5-8 GiB/s depending on the tensor size.
I'm aware of the UMD feature that @joelsmithTT is working on, and would like to try it when I get time (turns out we are reading device tensors via the fast dispatch patch which isn't going to get optimised with that feature, so we need some hacks to do to try it out properly)

Attaching a unit test which I used to measure read BW:
```
import time
import pytest
import torch
import ttnn

from models.utility_functions import (
    divup,
    profiler,
)

@pytest.mark.parametrize(
    "H, W",
    [
        (1024, 128),
        (2048, 128),
        (4096, 128),
    ],
    ids=["1024x128", "2048x128", "4096x128"],
)
def test_tensor_read(device, H, W):
    core_grid = device.compute_with_storage_grid_size()
    num_cores = core_grid.x * core_grid.y
    assert num_cores == 64

    C = 1
    torch_input_tensor = torch.randn(1, 1, C, H * W)
    ttnn_input_tensor = ttnn.from_torch(torch_input_tensor, dtype=ttnn.bfloat16)

    shard_width = H * W // num_cores
    shard_grid = ttnn.num_cores_to_corerangeset(num_cores, core_grid)
    shard_shape = [1, shard_width]

    shard_spec = ttnn.ShardSpec(
        shard_grid,
        shard_shape,
        ttnn.ShardOrientation.ROW_MAJOR,
    )
    l1_sharded_config = ttnn.MemoryConfig(
        ttnn.TensorMemoryLayout.WIDTH_SHARDED,
        ttnn.BufferType.L1,
        shard_spec
    )

    tensor_on_device = ttnn.to_device(ttnn_input_tensor, device=device, memory_config=l1_sharded_config)
    outputs = []

    print("Begin measurement")
    num_iters = 10000
    start = time.time()
    for i in range(num_iters):
        outputs.append(ttnn.from_device(tensor_on_device, blocking=False))
    ttnn.synchronize_device(device)
    end = time.time()

    total_time = end - start
    total_size_in_bytes = 2 * C * H * W * num_iters # 2 in for bfloat16
    total_size_in_GiB = total_size_in_bytes / (1024**3)
    time_per_tensor = total_time / num_iters
    print(f"Average time per tensor read for resolution: {H}x{W}: {time_per_tensor}s samples_per_sec: {1/time_per_tensor} GiB/s = {total_size_in_GiB / total_time}")
```
fyi @ttmtrajkovic @rjakovljevicTT @mbezuljTT @pavlejosipovic @pjanevskiTT

**Comment by pavlepopovic** [comment]
I've done a little bit more digging here - input tensor is huge, and therefore is actually presenting a bigger challenge to achieve our target performance on a customer model.
In short, our target for inference is ~244 us per tensor of size [1, 10, 2048, 128] (this is a bfloat16 tensor, so when calculating size we need to do a x2).
In order to hit that target we need to:
- optimize inference to be at least 244 us (we are ~3x off from this number at this time) (In samples/s, this number is roughly ~4100)
- optimize IO Writes and reads of the tensor.

Here is a breakdown of a test I've conducted on 3 machines (2xN150 (one with x16 and one with x8 PCIE), and 1 N300 (with x16 PCIE - on N300 we are using only L chip at the moment):

<img width="3084" alt="Image" src="https://github.com/user-attachments/assets/37c746bb-3b30-4d31-8e24-b5b1ea49ea61" />

In short:
Device inference samples/s is how many samples/s we currently have running on device (excluding IO Read/Write)
E2E perf run is a run of the model with trace + 2cq (input writes and reads are on CQ1, queued one after another (do they overlap??)
E2E run with model trace executed (only IO remaining) column is what happens if we have only input writes and output reads, queued one after another, without any inference happening.
It is obvious, that even if we manage to hit perf targets, we will be off by ~2x (on a N150 with x16) due to IO, hence this is a problem.
I also have an updated version of the test from above, which I will paste below, where we test read and write speeds to device, averaged over 10k iterations.
To run it, run:
pytest test_bw.py::test_tensor_read -k "2048" (add WH_ARCH_YAML in case of N300, to get 64 cores)
pytest test_bw.py::test_tensor_write -k "2048" (add WH_ARCH_YAML in case of N300, to get 64 cores)
One thing I've noticed is that, even with same PCIE width, N300 reads/writes are slower then N150? Is there any reason for that? The N300 CPU is a LoudBox configuration one.
What can we do to optimise IO reads/writes in this case?

```
import time
import pytest
import torch
import ttnn

from models.utility_functions import (
    divup,
)

@pytest.mark.parametrize(
    "H, W, C",
    [
        (1024, 128, 1),
        (2048, 128, 1),
        (4096, 128, 1),
    ],
    ids=["1024x128x1", "2048x128x1", "4096x128x1"],
)
def test_tensor_read(device, H, W, C):
    core_grid = device.compute_with_storage_grid_size()
    num_cores = core_grid.x * core_grid.y
    assert num_cores == 64

    torch_input_tensor = torch.randn(1, 1, C, H * W)
    ttnn_input_tensor = ttnn.from_torch(torch_input_tensor, dtype=ttnn.bfloat16)

    tensor_on_device = ttnn.to_device(ttnn_input_tensor, device=device, memory_config=ttnn.DRAM_MEMORY_CONFIG)
    outputs = []

    print("Begin measurement")
    num_iters = 10000
    start = time.time()
    for i in range(num_iters):
        outputs.append(ttnn.from_device(tensor_on_device, blocking=False))
    ttnn.synchronize_device(device)
    end = time.time()

    total_time = end - start
    total_size_in_bytes = 2 * C * H * W * num_iters # 2 in for bfloat16
    total_size_in_GiB = total_size_in_bytes / (1024**3)
    time_per_tensor = total_time / num_iters
    print(f"Average time per tensor read for resolution: {H}x{W}x{C}: {time_per_tensor}s samples_per_sec: {1/time_per_tensor} GiB/s = {total_size_in_GiB / total_time}")


@pytest.mark.parametrize(
    "H, W, C",
    [
        (1024, 128, 10),
        (2048, 128, 10),
        (4096, 128, 10),
    ],
    ids=["1024x128x10", "2048x128x10", "4096x128x10"],
)
def test_tensor_write(device, H, W, C):
    core_grid = device.compute_with_storage_grid_size()
    num_cores = core_grid.x * core_grid.y
    assert num_cores == 64

    torch_input_tensor = torch.randn(1, C, H, W)
    ttnn_input_tensor = ttnn.from_torch(torch_input_tensor, dtype=ttnn.bfloat16)

    dram_grid_size = device.dram_grid_size()
    dram_num_cores = dram_grid_size.x * dram_grid_size.y
    dram_shard_grid = ttnn.num_cores_to_corerangeset(dram_num_cores, dram_grid_size)

    dram_shard_shape = [
        divup(C * H, dram_grid_size.x),
        W,
    ]
    dram_shard_spec = ttnn.ShardSpec(
        dram_shard_grid,
        dram_shard_shape,
        ttnn.ShardOrientation.ROW_MAJOR,
    )
    dram_height_sharded_mem_config = ttnn.MemoryConfig(
        ttnn.TensorMemoryLayout.HEIGHT_SHARDED,
        ttnn.BufferType.DRAM,
        dram_shard_spec,
    )

    tensor_on_device = ttnn.to_device(ttnn_input_tensor, device=device, memory_config=dram_height_sharded_mem_config)

    print("Begin measurement")
    num_iters = 10000
    start = time.time()
    for i in range(num_iters):
        ttnn.copy_host_to_device_tensor(ttnn_input_tensor, tensor_on_device)
    ttnn.synchronize_device(device)
    end = time.time()

    total_time = end - start
    total_size_in_bytes = 2 * C * H * W * num_iters # 2 in for bfloat16
    total_size_in_GiB = total_size_in_bytes / (1024**3)
    time_per_tensor = total_time / num_iters
    print(f"Average time per tensor write for resolution: {H}x{W}x{C}: {time_per_tensor}s samples_per_sec: {1/time_per_tensor} GiB/s = {total_size_in_GiB / total_time}")

```

**Comment by pjanevskiTT** [comment]
@pavlepopovic can you try using Joel's https://github.com/tenstorrent/tt-umd/pull/735 DMA API for reading data from device if possible, see how results look like then

**Comment by esmalTT** [comment]
Adding UNet Shallow label since this is also blocking our 4k fps milestone.

**Comment by mbezuljTT** [comment]
Thanks @pgkeller for your input

re #2 is the number in C++ unit test 15GB/s /w PCIe x16 for host to device transfers? if yes, that is about 50% of utilization; and bellow what we need for a customer model to work (20GB/s, 70% utilization)

**Comment by pgkeller** [comment]
> Thanks [@pgkeller](https://github.com/pgkeller) for your input
> 
> re 2. is the number in C++ unit test 15GB/s /w PCIe x16 for host to device transfers? if yes, that is about 50% of utilization; and bellow what we need for a customer model to work (20GB/s, 70% utilization)

Yes, that's about right.  I think optimizing that comes down to not copying on the host plus potentially a dispatch optimization

**Comment by broskoTT** [comment]
Re 1.: Per https://tenstorrent.atlassian.net/browse/SYS-1045,  DMA host -> device works properly, but there are problems with the device->host direction.

Re 4.: Yes, this is a long term solution that we're trying to figure out how to properly support, that would avoid one host->host copy

**Comment by mbezuljTT** [comment]
Thank you @pgkeller 

3. DRAM interleaved, in the example above (that is torn-down from the model)
- we are reading (4096 * 128 * 2B) page size and only 1 bank
- we are writing 128 * 2B page size, and we have 10 channels thus 10 banks.  
If you know something would work better, please let us know.
4. it's overlapped, we're using 2xCQs & Trace. One CQ is for model, the other is for read/write. 
5. Agreed that we need to minize host copy, as far as I know @pjanevskiTT  was looking into it from the UMD side - like how to expose mapped buffer to the user. 
6. N300 scenario was using only left/local/MMIO chip

@esmalTT could you please share your scenario on Paul's questions 3 and 6 for shallow unet? that one is definitely using two chips.

@pgkeller I have a question: PCIe is full duplex, can we utilize overlapped reads and writes with 1 CQ? if not, does it suggest we might need 3 CQs? One for writes, one for inference, one for reads?

**Comment by pgkeller** [comment]
@mbezuljTT making sure I follow, page size is 4K*128*2B for 1 bank?  This is going to get stored in a single DRAM which is not efficient.  Ideal page sizes are 4K-8K.  Given the performance you are seeing, you are not hitting this bottleneck yet, but this is definitely not ideal.  Using 1 bank on WH will max out at about 22GB/s while inrerleaved across all DRAMs is closer to 28GB/s.

We cannot overlap read/write today, if we wanted to do this you are right, we'd need another CQ.  We can look into this.  My hunch is that we'll saturate NOC/DRAM BW before PCIe in this case and so the win wouldn't be 2x, but it could still be signficant.

However, I need to point out that the limits we are seeing from higher in the stack are much lower than the metal unit tests, so there is something going on there that needs to be addressed first.

**Comment by pjanevskiTT** [comment]
Wanted to give the update on UMD side as well.

https://github.com/tenstorrent/tt-umd/issues/751 tracks all the work and linked documents with the results of multiple measurements. Few key takeaways from the experiments

1. MMIO (TLB) path is simply too slow to be used in any perf path that we need. We are seeing ~4GB/s for writes and 5MB/s for reads. We are probably never going to use that in any perf paths
2. PCIE DMA case looks promising. We are seeing ~15GB/s for Tensix write and ~13GB/s for DRAM write e2e. This includes both memcpy to host memory that device can use for DMA transfers, as well as actual DMA transfer. I think this is close to e2e numbers metal is seeing when using NOC DMA for reading/writing using hugepages. As @pgkeller said

> not sure we'll reach the theoretical limit so long as we do a host  copy

main slowdown on this path is constantly copying data to host memory that we use for DMA, ideally we would want to avoid that. Today, I think metal can still try to use host memory in a smarter way in a way that they should use hugepages as area where to allocate tensors, commands and all the other stuff if possible. From then on, kernels should use that memory directly to read/write data. Other potential improvement comes as mapping already existing memory to be visible to device (point 3)
 
3. Mapping of host memory to be visible to the device. This is a feature that can be enabled when we turn the IOMMU on. It basically enables the application to map any existing memory to be visible to a device and then use the kernels to read/write data to this memory. However, operation of mapping this host buffers to a device visible address space also comes with a cost. I am currently measuring that in https://github.com/tenstorrent/tt-umd/pull/889, will come back with the results. I have already added tests which showcase how to use the potential new host memory scheme in https://github.com/tenstorrent/tt-umd/pull/857

To conclude, I think metal can already today try and use hugepages in a smarted way to avoind doing host copy. UMD measurements have shown that we are seeing **25GB/s transfer to Tensix and 21GB/s transfer to DRAM when PCIE DMA just reads/writes data without doing host copy**. Further improvements can come in a sense, that if metal can't avoid host copy at all times, UMD can provide the API to directly map some host memory to be visible to a device, from which kernels could read/write data. Let me know if you need any more details on this

for visibility @joelsmithTT @mbezuljTT @rjakovljevicTT @broskoTT

**Comment by smehtaTT** [comment]
Hi @mbezuljTT - this bug is blocking the Customer team and impacting 2 models (ResNet50 and Vit-Base).
Can we prioritize this to P0 and get an ETA on this?
FYI @mbahnasTT @mventerTT

**Comment by mventerTT** [comment]
@rjakovljevicTT also FYI

**Comment by pgkeller** [comment]
assigned @jbaumanTT as work on this is in flight

**Comment by jbaumanTT** [comment]
We now support pinned memory on systems with the IOMMU enabled. When using copy_host_to_device_tensor and copy_device_to_host_tensor I'm seeing maximum speeds of around 21GB/s on blackhole and 19GB/s on wormhole, depending on the page size and other constraints. Hopefully we've done enough on this now.
